strawberryfields/backends/fockbackend/ops.py

```diff
@@ -318,6 +338,109 @@ def apply_gate_einsum(mat, state, pure, modes, n, trunc):
     return np.einsum(einstring, mat, state, mat.conj())
 
 
+def apply_twomode_gate(mat, state, pure, modes, n, trunc, gate="BSgate"):
+    """Applies a two-mode gate to a state
+
+    Applies the two-mode gate to the state using custom tensor contractions and
+    the numba compiler for faster application.
+
+    Args:
+        mat (ndarray): The BS operator to be applied to the state
+        state (ndarray): The state that the BS is applied to
+        pure (bool): If the state is pure or mixed
+        modes (list[int]): A list of modes to which the BS is applied
+        n (int): The total number of modes
+        trunc (int): The Hilbert space truncation/cutoff
+        gate (str): the gate which should be called (BSgate, S2gate)
+
+    Returns:
+        ndarray: State where the two-mode operation has been applied
+    """
+    if pure:
+        t1 = modes[0]
+        t2 = modes[1]
+
+        # put the ket-values in front to be operated on in the apply function
+        switch_list_1 = np.arange(n)
+        switch_list_2 = np.arange(n)
+        switch_list_1[[0, t1]] = switch_list_1[[t1, 0]]
+        switch_list_2[[1, t2]] = switch_list_2[[t2, 1]]
+
+        state = state.transpose(switch_list_1)
+        state = state.transpose(switch_list_2)
+
+        if gate == "BSgate":
+            state = _apply_BS(mat, state, trunc)
+        elif gate == "S2gate":
+            state = _apply_S2(mat, state, trunc)
+        else:
+            raise NotImplementedError
```

Comment:
Good practice to specify why the error is raised:
```suggestion
            raise NotImplementedError("Currently, selection rules are only implemented for the BSgate "
                                      "and the S2gate. The {} gate is not supported".format(gate)")
```